docs/smart_bits.md

```diff
@@ -0,0 +1,71 @@
+
+The bit-width of any partial expression's result is determined by the specific operator (see table below). For example the result of addition has bit-width equal to the maximum of all of it's operand bit-widths, where as the result of comparison always has bit-width equal to 1.
+
+### Signedness
+The signedness of operands can both determine the semantics of operators, as well as how operands are extended. For example, signed-comparison has different semantics than unsigned-comparison, and operands can either be sign-extended or zero-extended (during bit-width inference).
+
+The rule of thumb for both self-determined and context-determined expressions is that all it's operands must be signed for the operator to be interpreted as signed (and therefore produce a signed output). There are the following exceptions:
+* The result of concat is always unsigned
+* The result of comparisons is always unsigned
+
+Additionally, there are two conversion functions (`signed()`, `unsigned()`) which coerce results of partial expressions to be either signed or unsigned, respectively. Note that these functions **do not** affect the signedness of any operands of the partial expression being coerced, but do affect the signedness of any other expressions in which the partial expression appears. For example, in `x + signed(y + z)` the signedness of `y` and `z` is not affected, but `(y + z)` is interpreted as a signed value in the expression `x + (y + z)`.
+
+### Summary of bit-width and signedness rules
+| Expression             | Operands           | Output bit-width | Output signedness       |
+|------------------------|--------------------|------------------|-------------------------|
+| Binary arith.,logical  | Context-determined | max(operands)  | all(signed(operands)) |
+| Unary logical          | Context-determined | max(operands)  | all(signed(operands)) |
+| Shift                  | Shiftee context-determined, shifter self-determined, | Shiftee bit-width | all(signed(operands)) |
+| Comparison             | Context-determined | 1                | unsigned                |
+| Reduction              | Self-determined    | 1                | all(signed(operands)) |
+| Concat                 | Self-determined    | sum(operands)  | unsigned |
+
+## TODOs
+* Integer operands
+* Explicit extend operator
+* SmartBits select operator
+* Signedness for SmartBit
+
+## Further reading
+* [Bit-width and signedness gotchas](http://www.deepchip.com/items/0466-05.html)
+
```

Comment:
Not for this PR, but in the future, it would be interesting to add some sort of context manager pattern for entering "SmartBits" syntax regions, where any incoming variables are converted to smartbits (this would make it convenient for interfaces with standard bits, but easily converting inputs into smart bits without having to explicitly case all of them by name).

Reply:
Yes, something like that would be super useful. I think in general (a) how we interoperate with non-smart code will be important for adoption and backwards compatibility, and (b) avoiding the use of "smart" literally in the code everywhere will be useful. Added a note about this to the docs TODO.